### PR TITLE
Permission extensions

### LINKED ISSUE #157: Potential incorrect WPS-1 service ACL permissions validation
see crim-ca/weaver#22 for use case

- [ ] Verify that `POST` for WPS-1 `request=Execute` doesn't cause problem (when using XML body instead of `DataInputs=...` in URI with `GET` request). When request has XML body, the values inside it are used instead of `identifier=...&request=...` in the URI. Therefore, the `ServiceWPS` possibly fails to retrieve the actual  `request` field, and then fails to retrieve the proper ACL for Twitcher.
- [ ] verify that WPS-1 `request=...` to corresponding Magpie service permission mapping is properly accomplished so that access to corresponding requests is granted/blocked as expected.

### LINKED ISSUE #234: WMS parsing of THREDDS service
Support the new functionality of integrated WMS in Thredds.

### LINKED ISSUE #361: [Feature] BROWSE permission for ServiceTHREDDS
- [x] add `BROWSE`permission to `THREDDS`'s `Directory` to allow listing of *catalog* contents. 
    `READ` becomes relative only to `Files` (although `Directory` can have it for `RECURSIVE` resolution, but does not employ it by itself. 
- [x] update doc about new behaviour

see details: 
- https://github.com/Ouranosinc/Magpie/pull/353#issuecomment-711079626
- https://github.com/Ouranosinc/Magpie/pull/353#issuecomment-710495259

### PR DESCRIPTION
**Pull request by fmigneault**
Depends on #353 (rebased). 

- Adds additional `BROWSE` permission to `THEDDS` service (#361)
- Adds many doc details about service implementations.
- Fix and adds test cases for #157 
- Fixes and test some edge case that failed to grant access when target `Resource` was missing (not explicitly defined in db), but that parent `Resource` had recursive-scope `Permission`.

Fixes #157 
Fixes #234
Fixes #361

### CODE REVIEW

File: docs/services.rst [hunk 1]
```diff
@@ -15,78 +19,384 @@ Each :term:`Service` is a specialization of a :term:`Resource` (see: :class:`mag
...
 Service Synchronization
 ------------------------
 
-.. todo:
-    resource auto-sync feature for a given service and cron configuration setup for it
+.. versionadded:: 0.7
+
+Some :term:`Service` implementations offer a synchronization feature which is represented by field ``sync_type`` within
+`Magpie` API responses. When this option is available, the corresponding :term:`Service` is able to query the real
+*remote service provider* to retrieve actual :term:`Resource` nested under it. This allows quick generation of the full
```
Comment (dbyrns) [inline docs/services.rst:384]: I think that the tree hierarchy is not fully recovered... only at a given depth.

File: docs/services.rst [hunk 2]
```diff
@@ -15,78 +19,384 @@ Each :term:`Service` is a specialization of a :term:`Resource` (see: :class:`mag
...
+
+When using the `Magpie` Docker image, the default command run the `magpie-cron`_ utility in parallel to the API. This
+cron job will periodically execute the :term:`Resource` auto-synchronization feature for a given :term:`Service` that
+supports it.
 
 .. seealso::
 
-    Utility ``magpie_sync_resources`` in `Magpie CLI Helpers`_ is also available to manually launch a :term:`Resource`
-    synchronization operation for supporting :term:`Service`-types.
+    Utility ``magpie_sync_resources`` in :ref:`utilities_helpers` is also available to manually launch a
```
Comment (dbyrns) [inline docs/services.rst:400]: Maybe refer also to the ui synch button...

File: magpie/services.py [hunk 3]
```diff
@@ -672,48 +681,64 @@ class ServiceTHREDDS(ServiceInterface):
...
+    def get_config(self):
+        # type: () -> ConfigDict
+        cfg = super(ServiceTHREDDS, self).get_config()
+        cfg.setdefault("file_patterns", [r".*.nc"])
+        cfg.setdefault("data_type", {"prefixes": []})
+        if not cfg["data_type"]["prefixes"]:
+            cfg["data_type"]["prefixes"] = ["fileServer", "dodsC", "dap4", "wcs", "wms"]
+        cfg.setdefault("metadata_type", {"prefixes": []})
+        if not cfg["metadata_type"]["prefixes"]:
+            cfg["metadata_type"]["prefixes"] = [None, "catalog", "ncml", "uddc", "iso"]
```
Comment (dbyrns) [inline magpie/services.py:693]: You provide a great default configuration, but to avoid a back-compatibility migration we could also push all prefixes to the data type. The responsability of the admin overriding this config with metadata prefixes would then have to set the permissions of his instance accordingly.
Note that I would prefer a proper migration, but understand that it could be complex.

File: magpie/services.py [hunk 4]
```diff
@@ -672,48 +681,64 @@ class ServiceTHREDDS(ServiceInterface):
...
-        while found_resource and path_parts:
+        found_resource = child_resource = self.service
+        while child_resource and path_parts:
             part_name = path_parts.pop(0)
-            found_res_id = found_resource.resource_id
-            found_resource = models.find_children_by_name(part_name, parent_id=found_res_id, db_session=self.request.db)
+            # when reaching the final part, test for possible file pattern, otherwise default to literal value
+            #   allows combining different naming formats into a common file resource (eg: extra extensions)
+            # if final part is a directory, still works because of literal value
+            #   directory name much match exactly, no format naming variants allowed
```
Comment (dbyrns) [inline magpie/services.py:712]: typo "much" -> "must"

**Review by dbyrns — CHANGES_REQUESTED** [review]
Good job! I only got some small questions and suggestions.

File: docs/services.rst [hunk 5]
```diff
@@ -15,78 +19,384 @@ Each :term:`Service` is a specialization of a :term:`Resource` (see: :class:`mag
...
+The implementation of this :term:`Service` is handled by class :class:`magpie.services.ServiceAccess`. It is intended to
+act as a simple *all-or-nothing* endpoint barrier, where only :attr:`Permission.ACCESS` can be applied, and only
+directly on the :term:`Service` itself. A :term:`User` or :term:`Group` that has that :term:`Permission` with
+:attr:`Access.ALLOW` will be able to reach the :term:`Service`. Any other operation (or explicit :attr:`Access.DENY`)
+will all result into denied access to the private URL registered by the :term:`Service`.
+
+.. versionchanged:: 3.0
+    This :term:`Service` implementation dates prior to the integration of :class:`Access` and :class:`Scope` concepts
+    that now provides modifiers to the :term:`Permission` resolution methodology, notably prior to the addition of
+    :attr:`Access.DENY`. This :term:`Service` is equivalent to a simplified version of the fine-grained `ServiceAPI`_
```
[the remark this answers is not in the document]
Comment (fmigneault) [reply docs/services.rst:114]: Could be. I leave it for now because it is not really painful to maintain.

File: docs/services.rst [hunk 6]
```diff
@@ -15,78 +19,384 @@ Each :term:`Service` is a specialization of a :term:`Resource` (see: :class:`mag
...
+    The :attr:`Permission.BROWSE` permission is used to provide listing access of contents when targeting a
+    :term:`Resource` of type :class:`magpie.models.Directory`. When targeting a :class:`magpie.models.File`, it instead
+    provides *metadata* access to that file.
+
+Permission :attr:`Permission.READ` can be applied to all of the resources, but will only effectively make sense when
+attempting access of a specific :term:`Resource of type :class:`magpie.models.File`.
+
+.. versionchanged:: 3.1
+    Permission :attr:`Permission.READ` does not offer *metadata* content listing of :class:`magpie.models.Directory`
+    anymore. For this, :attr:`Permission.BROWSE` should be used instead. Setting :attr:`Permission.READ` on a
```
[the remark this answers is not in the document]
Comment (fmigneault) [reply docs/services.rst:176]: Nothing is done for the moment, but we could had this step to the db migration scripts.
That's a good point actually. Users can only download but not query metadata at the moment if READ only.

File: docs/services.rst [hunk 7]
```diff
@@ -15,78 +19,384 @@ Each :term:`Service` is a specialization of a :term:`Resource` (see: :class:`mag
...
+:term:`Effective Permissions` resolution, the value of the query parameter ``identifier`` is employ to attempt mapping
+it against an existing :term:`Resource`. The resolution of :term:`Effective Permissions` in even of multi-level tree
+:term:`Resource` is computed in the usual manner described in the :ref:`Permissions` chapter.
+
 
+.. warning::
+    When applying permissions on a per-:class:`magpie.models.Process` basis, :attr:`Scope.MATCH` modifier is recommended
+    if only a specific ``request`` should be granted access, although :attr:`Scope.RECURSIVE` would have the same effect
+    currently because these resources are necessarily leaf nodes by definition. This is to prevent unexpectedly granting
+    additional lower-level :term:`Resource` access in the event this definition gets modified or extended in the future.
```
[the remark this answers is not in the document]
Comment (fmigneault) [reply docs/services.rst:365]: The Service and the children Resource can have very distinct permission names if we want. 
I actually use this for `ServiceWPS` vs `Process` under it. Only `ServiceWPS` can have `GET_CAPABILITIES`, and both can have `DESCRIBE_PROCESS` and `EXECUTE`.

File: docs/services.rst (hunk 1 above)
Comment (fmigneault) [reply docs/services.rst:384]: Specifically for THREDDS, there is an hardcoded `max_depth = 3`. Others return `None` which will fetch everything. 
Good catch.

File: docs/services.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (dbyrns) [reply docs/services.rst:365]: Hum, that's good! You just cannot restrict MATCH over RECURSIVE to avoid the potential issue listed above.

File: magpie/services.py (hunk 3 above)
Comment (fmigneault) [reply magpie/services.py:693]: Indeed. There is quite a bit of work to move this to the db properly. 
Saving the config the db in itself is somewhat easy, but the problem is about how to handle updating it. 

Example: 
First time the config is provided via config.yml and stored to database. 
Then, if omitted another time, how do we handle this? Reuse the thing from db or flush? How to know what is the intended behavior? The only real method would be to have an API call to update, but then more development to support this.

For now, I think it is safe to assume that the config.yml provided will always be there, so the config should be loaded each time as expected.

Comment (dbyrns) [reply magpie/services.py:693]: Hum I'm not sure you understand. I don't want to save the config to the db.
I only point out that for an instance without proper thredds configuration (all of them as of speaking), when we will update magpie, the ``if not cfg[...][...]`` will be true and the "default" behavior will apply. This default behavior, with the current instances, will cause to remove all permissions for the metadata_type [None, "catalog", "ncml", "uddc", "iso"] since no BROWSE permissions exist. 
What I propose here, is a solution to https://github.com/Ouranosinc/Magpie/pull/363#discussion_r512801739. If by default, you put all the prefix in ``cfg["data_type"]["prefixes"]``, then when upgrading magpie, even without BROWSE permissions set, everything will work as before. 
The admin will then be free to configure the browse permission for its users/groups and later configure thredds properly using the config.yml.

Comment (fmigneault) [reply magpie/services.py:693]: Although that would work, it somewhat goes against to point of adding the permission specifically for metadata in the first place. Isn't it be better to have a default config that matches as closely as possible "normal" usage? I would argue that if for whatever reason we want to preserve the old method, we should provide the config to do so.

When Magpie is going to be updated to support this specific use case, proper configuration should be provided to it.
I think it is cleaner to duplicate existing READ entries as equivalent BROWSE during migration.

Comment (dbyrns) [reply magpie/services.py:693]: Yes I agree, that's why I noted that I would prefer a proper migration. I just tought that it may be hard, but let's do that. Assign BROWSE permissions where READ permissions currently exist.

File: docs/services.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (fmigneault) [reply docs/services.rst:176]: added by d364884

File: docs/services.rst [hunk 8]
```diff
@@ -15,78 +19,384 @@ Each :term:`Service` is a specialization of a :term:`Resource` (see: :class:`mag
...
+permission will be immediately assumed as :attr:`Access.DENY` regardless of type. To allow top-level access directly on
+the :term:`Service`'s root without ``<prefix>``, it is important to provide ``null`` within the desired ``prefixes``
+list. Duplicates between the two lists of ``prefixes`` will favor entries in ``metadata_type`` over ``data_type``.
+
+After resolution of the content type from ``<prefix>``, the resolution of any amount of :class:`magpie.models.Directory`
+:term:`Resource` will be attempted. Any missing children directory :term:`Resource` will terminate the lookup process
+immediately, and :term:`ACL` will be resolved considering :attr:`Scope.RECURSIVE` if any applicable parent
+:term:`Resources` for the given :term:`Permission` selected by ``<prefix>`` and from where lookup stopped.
+
+Once the last element of the path is reached, the ``file_patterns`` will be applied against ``<file>`` in order to
```
[the remark this answers is not in the document]
Comment (fmigneault) [reply docs/services.rst:260]: added by 5cabd1b

File: docs/services.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (fmigneault) [reply docs/services.rst:375]: Yes. Thus the warning.
It actually applies to any resource (that allows sub-children) that is located at the lowest *current* level of the tree, regardless of service/resource-type. Adding recursive allow access to resources that could be added at a later time under it.

**Review by dbyrns — APPROVED** [review]
Everything's fine with me! Thank you.

**Comment by fmigneault** [comment]
@dbyrns redeployed on ogc-ades
